File: antia_numerical_methods/f77/xmp4/sum.f
!     PROGRAM TO SUM A FINITE SERIES USING CASCADE SUM ALGORITHM
!     TERM IS SPECIFIED AS A FUNCTION

      PROGRAM SUMSER
!      IMPLICIT REAL*8(A-H,O-Z)
      EXTERNAL TERM

!     EXAMPLE 2.2

51    FORMAT('  N =',I8,5X,'CASCADE SUM =',1PE14.6)

100   PRINT *,'TYPE N = NUMBER OF TERMS (QUITS WHEN N.LE.0)'
      READ *,N
      IF(N.LE.0) STOP
      S=CASSUM(TERM,N)
      WRITE(6,51) N,S
      GO TO 100
      END
 
!     ---------------------------------------------------------
 
!	To find cascade sum of a series
!
!	TERM : (input) Name of function routine to calculate the Ith term
!	      It is possible to supply the terms in a real array TERM
!	      if the dimension statement is uncommented. (see CASSUM_A)
!	N : (input) Number of terms to be summed
!
!	FUNCTION TERM(I) to calculate the Ith term must be supplied by
!	the user
!
!	Required routines : TERM

      FUNCTION CASSUM(TERM,N)
!      IMPLICIT REAL*8(A-H,O,P,R-Z)
      IMPLICIT LOGICAL(Q)
      PARAMETER(N2MAX=30)
      DIMENSION S(N2MAX),Q(N2MAX)
!	Uncomment this line if TERM is required to be an array 
!	DIMENSION TERM(N)

      CASSUM=0.0
      DO 1000 I=1,N2MAX
1000  Q(I)=.TRUE.

      DO 2000 I=1,N
        IF(Q(1)) THEN
          S(1)=TERM(I)
          Q(1)=.FALSE.

!	If a pair is formed add the sum to higher level in the binary tree
        ELSE
          S(1)=S(1)+TERM(I)
          Q(1)=.TRUE.
          DO 1500 J=2,N2MAX
            IF(Q(J)) THEN
              S(J)=S(J-1)
              Q(J)=.FALSE.
              GO TO 2000
            ELSE
              S(J)=S(J)+S(J-1)
              Q(J)=.TRUE.
            ENDIF
1500      CONTINUE

          CASSUM=CASSUM+S(N2MAX)
        ENDIF

2000  CONTINUE

!	Find the sum by adding up the incomplete pairs
      DO 3000 J=1,N2MAX-1
        IF(.NOT.Q(J)) THEN
          IF(Q(J+1)) THEN
            S(J+1)=S(J)
            Q(J+1)=.FALSE.
          ELSE
            S(J+1)=S(J+1)+S(J)
          ENDIF
        ENDIF
3000  CONTINUE
      IF(.NOT.Q(N2MAX)) CASSUM=CASSUM+S(N2MAX)
      END
 
!     --------------------------------------------------
 
      FUNCTION TERM(N)
!      IMPLICIT REAL*8(A-H,O-Z)

!     N TH TERM OF THE REQUIRED SERIES

      TERM=(-1.0)**(N+1)/N
      END

File: projects/antia_numerical_methods@Beliavsky/c/xmp/sum.c
/*	Cascade sum of a finite series */

#include <stdio.h>
#include <math.h>

double cassum(double (*term)(int ), int n);
double terms(int i);

main()
{
	int i,j,n;
	double s;

/*	Example 2.2 */

	for(i=0; i<99; ++i) {
		printf("type n = no. of terms   (quits when n<=0) \n");
		scanf("%d", &n);
		if(n<=0) return 0;
		s=cassum(terms,n);
		printf(" n = %d   sum = %f \n", n,s);
	}
	return;
}

 

/*	To find cascade sum of a series

	TERM : (input) Name of function to calculate the Ith term
	      It is possible to supply the terms in an array TERM
	      if TERM is declared as an array. (see CASSUM_A)
	N : (input) Number of terms to be summed

	Function TERM(I) to calculate the Ith term must be supplied by
	the user

	Required functions : TERM
*/

#include <math.h>


double cassum(double (*term)(int ), int n)

{
	int i, j, q[30], n2max=30;
	double s[30], sum;
	enum boolean {NO, YES};

	sum=0.0;
	for(i=0; i<n2max; ++i)  q[i]=YES;

	for(i=0; i<n; ++i) {
		if(q[0]) {s[0]=term(i); q[0]=NO;}
/*	If a pair is formed add the sum to higher level in the binary tree */
		else {
			s[0]=s[0]+term(i); q[0]=YES;
			for(j=1; j<n2max; ++j) {
				if(q[j]) {s[j]=s[j-1]; q[j]=NO; break;}
				else {s[j]=s[j]+s[j-1]; q[j]=YES;}
			}
			if(j>=n2max) sum=sum+s[n2max-1];
		}
	}

/*	Find the sum by adding up the incomplete pairs */

	for(j=0; j<n2max-1; ++j) {
		if(q[j]==NO) {
			if(q[j+1]) {s[j+1]=s[j]; q[j+1]=NO;}
			else {s[j+1]=s[j+1]+s[j];}
		}
	}
	if(q[n2max-1]==NO) sum=sum+s[n2max-1];
	return sum;
}



double terms(int i)

{
	double t;
	t=1./(i+1.);
	if(i-2*(i/2) == 1) t=-t;
	return t;
}
